README.md
# Meadow — WordPress Templating DSL

_Write WordPress theme templates with familiar ease and modern features._

[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/Rarst/meadow/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/Rarst/meadow/?branch=master)
[![Version](https://img.shields.io/packagist/v/rarst/meadow.svg?label=version)](https://packagist.org/packages/rarst/meadow)
[![PHP required](https://img.shields.io/packagist/php-v/rarst/meadow.svg)](https://packagist.org/packages/rarst/meadow)
[![PDS Skeleton](https://img.shields.io/badge/pds-skeleton-blue.svg?style=flat-square)](https://github.com/php-pds/skeleton)

Meadow is a theme templating solution, aiming to find a balance between native WordPress concepts and power of [Twig](http://twig.sensiolabs.org/) dedicated templating language.

## Installation

Require package in your theme project with [Composer](https://getcomposer.org/):

```bash
composer require rarst/meadow
```

Instantiate object some time during theme load:

```php
$meadow = new \Rarst\Meadow\Core;
$meadow->enable();
```

## Templating

Meadow follows conventions of WordPress [template hierarchy](https://codex.wordpress.org/Template_Hierarchy#Visual_Overview):

 - for example `index.php` becomes `index.twig`.
 - `{{ get_header() }}` will look for `header.twig` (with fallback to `header.php`)
 - and so on.

### Template Tags

Template Tags API (and PHP functions in general) are set up to work transparently from Twig templates:

```twig
{{ the_title() }}
```

### Filters

WordPress filters set up to be available as Twig filters:

```twig
{{ 'This is the title'|the_title }}
```

### Template Inheritance

Full range of Twig functionality is naturally available, including [template inheritance](http://twig.sensiolabs.org/doc/templates.html#template-inheritance):

```twig
{# single.twig #}
{% extends 'index.twig' %}

{% block entry_title %}
	<div class="page-header">{{ parent() }}</div>
{% endblock %}
```

To inherit parent template in child theme prepend it with folder's name:

```twig
{# child-theme/index.twig #}
{% extends 'parent-theme/index.twig' %}
```

## Domain Specific Language

Meadow attempts not just "map" WordPress to Twig, but also meaningfully extend both to improve historically clunky WP constructs.

This is primarily achieved by implementing custom Twig tags, abstracting away complexities for specific tasks.

### Loop

```twig
{% loop %}
	<h2><a href="{{ the_permalink() }}">{{ the_title() }}</a></h2>
	{{ the_content() }}
{% endloop %}
```

### Secondary Loop

```twig
{% loop { 'post_type' : 'book', 'orderby' : 'title' } %} {# expression for arguments #}
	<h2><a href="{{ the_permalink() }}">{{ the_title() }}</a></h2>
	{{ the_content() }}
{% endloop %}
```

### Comments

```twig
<ul class="comment-list">
	{% comments %}
	<li>
		{{ comment_text() }}
	{# no </li> - self-closing #}
	{% endcomments %}
</ul>
```

## Template Examples

In [Hybrid Wing](https://github.com/Rarst/hybrid-wing) theme (work in progress):

 - [`index.twig`](https://github.com/Rarst/hybrid-wing/blob/master/index.twig)
  - [`single.twig`](https://github.com/Rarst/hybrid-wing/blob/master/single.twig)
   - [`single-post.twig`](https://github.com/Rarst/hybrid-wing/blob/master/single-post.twig)
  - [`comments.twig`](https://github.com/Rarst/hybrid-wing/blob/master/comments.twig)

## License

MIT

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

## 0.3 - 2018-12-31

### Changed
- code structure for PDS and PSR-4.
- dependencies (Twig 2+, PHP 7+).

### Fixed
- possible double output in `template_include`.

## 0.2 - 2016-12-10

### Added
- new hierarchy implementation for WP 4.7+ 

### Deprecated
- old hierarchy implementation for WP <4.7, to be removed in 1.0

### Fixed
- deprecated methods in Twig extension

## 0.1 - 2015-04-13

src/Template_Hierarchy.php
<?php
namespace Rarst\Meadow;

/**
 * Augment native template hierarchy with non-PHP template processing.
 *
 * @deprecated 0.2:1.0 Deprecated in favor of better hierarchy filter in WP 4.7+.
 */
class Template_Hierarchy {

	/** @var string[] $template_types Template type names to be used for dynamic hooks. */
	public $template_types = array(
		'embed',
		'404',
		'search',
		'taxonomy',
		'frontpage',
		'home',
		'attachment',
		'single',
		'page',
		'singular',
		'category',
		'tag',
		'author',
		'date',
		'archive',
		'commentspopup',
		'paged',
		'index',
	);

	protected $mime_type;

	public function enable() {

		add_action( 'template_redirect', array( $this, 'template_redirect' ) );

		foreach ( $this->template_types as $type ) {
			add_filter( "{$type}_template", array( $this, 'query_template' ) );
		}
	}

	public function disable() {

		remove_action( 'template_redirect', array( $this, 'template_redirect' ) );

		foreach ( $this->template_types as $type ) {
			remove_filter( "{$type}_template", array( $this, 'query_template' ) );
		}

		if ( ! empty($this->mime_type) ) {
			remove_filter( "{$this->mime_type[0]}_template", array( $this, 'query_template' ) );
			remove_filter( "{$this->mime_type[1]}_template", array( $this, 'query_template' ) );
			remove_filter( "{$this->mime_type[0]}{$this->mime_type[1]}_template", array( $this, 'query_template' ) );
		}
	}

	public function template_redirect() {

		if ( is_attachment() ) {
			global $posts;

			if ( ! empty( $posts ) && isset( $posts[0]->post_mime_type ) ) {
				$this->mime_type = explode( '/', $posts[0]->post_mime_type );
			}

			add_filter( "{$this->mime_type[0]}_template", array( $this, 'query_template' ) );
			add_filter( "{$this->mime_type[1]}_template", array( $this, 'query_template' ) );
			add_filter( "{$this->mime_type[0]}{$this->mime_type[1]}_template", array( $this, 'query_template' ) );
		}
	}

	/**
	 * @param string $fallback
	 *
	 * @return string
	 */
	public function query_template( $fallback ) {

		$type      = substr( current_filter(), 0, - 9 ); // trim '_template' from end
		$templates = array();

		switch ( $type ) {
			case 'embed':

				$object = get_queried_object();

				if ( ! empty( $object->post_type ) ) {

					$post_format = get_post_format( $object );

					if ( $post_format ) {
						$templates[] = "embed-{$object->post_type}-{$post_format}.twig";
					}

					$templates[] = "embed-{$object->post_type}.twig";
				}

				$templates[] = 'embed.twig';

				break;

			case 'taxonomy':
				$term = get_queried_object();

				if ( $term ) {
					$taxonomy    = $term->taxonomy;
					$templates[] = "taxonomy-{$taxonomy}-{$term->slug}.twig";
					$templates[] = "taxonomy-{$taxonomy}.twig";
				}

				$templates[] = 'taxonomy.twig';
				break;

			case 'frontpage':
				$templates = array( 'front-page.twig' );
				break;

			case 'home':
				$templates = array( 'home.twig', 'index.twig' );
				break;

			case 'single':
				$object = get_queried_object();

				if ( $object ) {
					$templates[] = "single-{$object->post_type}.twig";
				}

				$templates[] = 'single.twig';
				break;

			case 'page':
				$page_id  = get_queried_object_id();
//				$template = get_page_template_slug();
				$pagename = get_query_var( 'pagename' );

				if ( ! $pagename && $page_id ) {
					// If a static page is set as the front page, $pagename will not be set. Retrieve it from the queried object
					$post     = get_queried_object();
					$pagename = $post->post_name;
				}

				// TODO page templates
//				if ( $template && 0 === validate_file( $template ) )
//					$templates[] = $template;

				if ( $pagename ) {
					$templates[] = "page-{$pagename}.twig";
				}

				if ( $page_id ) {
					$templates[] = "page-{$page_id}.twig";
				}

				$templates[] = 'page.twig';
				break;

			case 'category':
			case 'tag':
				$term = get_queried_object();

				if ( $term ) {
					$templates[] = "{$type}-{$term->slug}.twig";
					$templates[] = "{$type}-{$term->term_id}.twig";
				}

				$templates[] = "{$type}.twig";
				break;

			case 'author':
				$author = get_queried_object();

				if ( $author ) {
					$templates[] = "author-{$author->user_nicename}.twig";
					$templates[] = "author-{$author->ID}.twig";
				}

				$templates[] = 'author.twig';
				break;

			case 'archive':
				$post_types = array_filter( (array) get_query_var( 'post_type' ) );

				if ( \count( $post_types ) === 1 ) {
					$post_type   = reset( $post_types );
					$templates[] = "archive-{$post_type}.twig";
				}

				$templates[] = 'archive.twig';
				break;

			default:
				$templates = array( "{$type}.twig" );
		}

		$template = $this->locate_template( $templates );

		if ( empty( $template ) ) {
			$template = $fallback;
		}

		return apply_filters( 'meadow_query_template', $template, $type );
	}

	/**
	 * Broken out for easier inheritance to customize lookup logic.
	 *
	 * @param array|string $templates
	 *
	 * @return string
	 */
	public function locate_template( $templates ) {

		return locate_template( $templates );
	}
}
